### ISSUE TITLE
Multiple yAxis & calculated height yields wrong axis range

### ISSUE THREAD

**Issue by jfoshee** [highslide-software/highcharts.com#4442]
See: http://jsfiddle.net/aq9n8m1x/5/

Where there are 2 Y axes and the vertical height of the `div` is calculated the range of the Y axis is wrong.  The options `endOnTick` and `maxPadding` are not honored, and indeed make the issue more apparent. (See <http://jsfiddle.net/aq9n8m1x/3/>) 

The impact is that a lot of vertical space can be wasted by the chart. It depends on the data and window size.

`<div id="container" style="min-width: 310px; height: calc(100vh);"></div>`

```
$(function () {
    $('#container').highcharts({
        yAxis: [{}, {}],
        series: [{
            data: [991, 455],
            // BUG: Change yAxis to 0 to see correct behavior
            yAxis: 1
        }]
    });
});
```

**Comment by highslide-software** [comment]
With multiple axes, the axes will become aligned as per the [http://api.highcharts.com#chart.alignTicks](alignTicks) option. When there are multiple axes they go into a mode of calculating a fixed amount of ticks,  which makes the scaling more rigid than if it were just a single axis.

This isn't really necessary when the first axis is empty, so there's something wrong with this block:
```js
			// Check if there are multiple axes in the same pane
			each(this.chart[this.coll], function (axis) {
				var options = axis.options,
					horiz = axis.horiz,
					key = [horiz ? options.left : options.top, horiz ? options.width : options.height, options.pane].join(',');
				
				if (others[key]) {
					if (axis.series.length) {
						hasOther = true; // #4201
					}
				} else {
					others[key] = 1;
				}
			});
```

**Comment by pawelfus** [comment]
Note: test also against #4201

**Comment by jfoshee** [comment]
I'm delighted to see my test case was expanded to build the unit test for the issue! Thanks, @pawelfus!

**Comment by pawelfus** [comment]
Another test case when it fails: http://jsfiddle.net/aq9n8m1x/6/

Checking if axis contains series is not enough, in fact we should check if any of the series connected to the axis contains data (or rather has extremes?).

**Comment by pawelfus** [comment]
Better example: http://jsfiddle.net/aq9n8m1x/7/

**Comment by pawelfus** [comment]
We can not check `axis.dataMin/Max` / `axis.hasVisibleSeries` / `hasData()` - all that info fails, when rendering first yAxis, because the second one is undefined. Most probably we should loop over all series, and check manually if:

- any series exist
- at least one series is visible 
- at least one series contains points

Any other cases when we should set `hasOther flag to true?

**Comment by TorsteinHonsi** [comment]
I think those checks would be pretty safe. Should be added to `Axis.hasData()`?

**Comment by pawelfus** [comment]
I think so, `Axis.hasData()` should make this  extra checks if `axis.hasVisibleSeries` and/or `axis.dataMin/Max` are undefined? Then we could safely use `Axis.hasData()` when aligning ticks.

**Comment by TorsteinHonsi** [comment]
Yes, I think that would be a safe solution. You can also look at the `hasData` method in the `no-data-to-display.js` module, maybe we get some synergy there. But still focus on keeping the file size down in the main `highcharts.js` file.
